### PR TITLE
New method on RepositoryCommits client - Get the sha1 of a commit reference

### LINKED ISSUE #1189: New method on RepositoryCommits client
https://developer.github.com/v3/repos/commits/#get-the-sha-1-of-a-commit-reference

Added to the [API docs](https://github.com/github/developer.github.com/commit/524e1d748156a8cabfdb01a4ca112c7408bb24b8) on 27 February 2016

```
GET /repos/:owner/:repo/commits/:ref
```

Note this is a preview API and requires a custom accept header to activate the functionality
` application/vnd.github.chitauri-preview+sha`

### PR DESCRIPTION
**Pull request by rogertinsley**
**Title**
New method on RepositoryCommits client - Get the sha1 of a commit reference. Fixes #1189 

**Description**

- [x] Add a sha1 method to IRepositoryCommitsClient
- [x] Add unit and integration tests for this new method
- [x] Add a sha1 method to IObservableRepositoryCommitsClients
- [x] Update AcceptHeaders to include the preview http accept header

This is my first foray into contributing to OSS. Any feedback is greatly appreciated.

### CODE REVIEW

File: Octokit/Clients/IRepositoryCommitsClient.cs [hunk 1]
```diff
@@ -50,5 +50,14 @@ public interface IRepositoryCommitsClient
...
         Task<IReadOnlyList<GitHubCommit>> GetAll(string owner, string name, CommitRequest request);
+
+        /// <summary>
+        /// Get the SHA-1 of a commit reference
+        /// </summary>
+        /// <param name="owner">The owner of the repository</param>
+        /// <param name="name">The name of the repository</param>
+        /// <param name="reference">The repository reference</param>
+        /// <returns></returns>
+        Task<string> Sha1(string owner, string name, string reference);
```
Comment (shiftkey) [inline Octokit/Clients/IRepositoryCommitsClient.cs]: Could we add a verb to this method? Say something like `GetSha`?

**Comment by shiftkey** [comment]
@rogertinsley welcome!

To be perfectly honest, the only bits of feedback I have are minor - this is really nice work!

One suggestion - if you wanted to add another integration test where you:

 - create a repository
 - make a commit through the API, get the new SHA
 - verify the SHA from this new endpoint matches what you got earlier

I think that'll give us enough coverage for this endpoint to close this out.

**Comment by rogertinsley** [comment]
Thanks @shiftkey, I'll make the changes that you've suggested :+1:

File: Octokit.Tests/Clients/RepositoriesClientTests.cs [hunk 2]
```diff
@@ -790,5 +790,35 @@ public void GetsCorrectUrl()
                 await Assert.ThrowsAsync<ArgumentException>(() => client.EditBranch("owner", "repo", "", update));
             }
         }
+
+        public class TheSha1Method
+        {
+            [Fact]
+            public async Task EnsureNonNullArguments()
```
Comment (ryangribble) [inline Octokit.Tests/Clients/RepositoriesClientTests.cs]: Minor note that the wording should be "Ensures" rather than "Ensure".  Since the test shoudl read TheGetSha1Method.EnsuresNonNullArguments() :grinning:

Also this test is checking Null and Empty arguments.  In other places ive seen us implement this as 2 test methods eg `EnsuresNonNullArguments()` and `EnsuresNonEmptyArguments()`

**Comment by ryangribble** [comment]
Great first contribution... I've agree with @shiftkey regarding `GetSha1()` and the integration test... and I've also raised one minor nitpick on consistency/wording

**Comment by M-Zuber** [comment]
:clap: Nice stuff!
Welcome to OSS :smile:

**Comment by rogertinsley** [comment]
I've made the changes you both have suggested.

Are you happy with the integration test approach? I've used a similar approach to other tests in the same class, that is to use `CreateTheWorld` to create repo & commit. Get the Sha1 and compare that it hasn't changed.

File: Octokit.Tests/Clients/RepositoriesClientTests.cs [hunk 3]
```diff
@@ -790,5 +790,41 @@ public void GetsCorrectUrl()
                 await Assert.ThrowsAsync<ArgumentException>(() => client.EditBranch("owner", "repo", "", update));
             }
         }
+
+        public class TheSha1Method
+        {
+            [Fact]
+            public async void EnsuresNonNullArguments()
```
Comment (shiftkey) [inline Octokit.Tests/Clients/RepositoriesClientTests.cs]: Make this `async Task` as this is more test-runner friendly...

File: Octokit.Tests.Integration/Clients/RepositoryCommitsClientTests.cs [hunk 4]
```diff
@@ -158,6 +166,17 @@ public TestsWithNewRepository()
             Assert.Equal(0, result.BehindBy);
         }
 
+        [IntegrationTest]
+        public async Task GetSha1FromRepository()
+        {
+            await CreateTheWorld();
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/RepositoryCommitsClientTests.cs]: I'm not sure how much this will break, but I'd love to get the `Reference` that `CreateTheWorld` returns - it should contain a `Sha`.

So rather than invoking `GetSha1` twice below, you invoke it once and compare that SHA to the one you have here.

Does that make sense?

File: Octokit.Tests/Clients/RepositoriesClientTests.cs (hunk 3 above)
Comment (shiftkey) [inline Octokit.Tests/Clients/RepositoriesClientTests.cs]: And the CI test won't fail like it currently is...

**Comment by rogertinsley** [comment]
I've updated `CreateTheWorld` to return a `Task<Reference>` and compare it against the `GetSha1` and the class `TheSha1Method` methods should be async

File: Octokit.Tests/Clients/RepositoriesClientTests.cs [hunk 5]
```diff
@@ -790,5 +790,40 @@ public void GetsCorrectUrl()
                 await Assert.ThrowsAsync<ArgumentException>(() => client.EditBranch("owner", "repo", "", update));
             }
         }
+
+        public class TheSha1Method
```
Comment (ryangribble) [inline Octokit.Tests/Clients/RepositoriesClientTests.cs]: Should be `TheGetSha1Method`

Comment (rogertinsley) [inline Octokit.Tests/Clients/RepositoriesClientTests.cs]: Thanks - updated it, I'm getting there :smile:

**Comment by ryangribble** [comment]
This looks good!  

But I noticed there are no Unit or Integration tests for the `ObservableRepositoryCommitsClient`.

I realise this was like that before this PR... Some parts of the code base (particularly observable implementations) don't have tests around them, but as we touch these areas ideally we should start to correct that problem.  Without tests there really could be something lurking in there making the observable implementation not work and we dont want to find out from end users if possible!

So at a minimum, I'd like to see tests added for the Observable `GetSha1()` method that has been added... And ideally - tests for ALL the methods in this observable client could be added!  :grinning:

If you have a look at some existing examples of Observable client unit and integration tests you will find that the unit tests of observable would have a `CallsIntoClient` type test (that checks the underlying regular client method is called with appropriate args), as well as `EnsuresNonNullArguments` and `EnsuresNonEmptyArguments` tests as appropriate (depending on what parameters the call takes and whether they have checks for nulls/empty string).  Also a test for the `Ctor` guarding against a null client argument.

Im not sure where @shiftkey stands on integration tests for observable clients - given that alot are missing, and it's pretty much only ever calling the regular client anyway... perhaps all we need are unit tests for observable clients... 

But in my contributions I do typically add integration tests, I pretty much copy/paste the regular client test, modify it to use the observable client and also like to write it so it's obvious the item is an observable (eventhough you could simply `await` it inline).  This is just personal preference.

``` csharp
var observable = _fixture.GetSha1(x,x,x);
var sha1 = await observable;
```

**Comment by rogertinsley** [comment]
I've added unit and integrations tests for the `ObservableRepositoryCommitsClient`, specifically the `GetSha1` method.

**Comment by shiftkey** [comment]
> Im not sure where @shiftkey stands on integration tests for observable clients - given that alot are missing, and it's pretty much only ever calling the regular client anyway... perhaps all we need are unit tests for observable clients...

This is essentially my views - if the implementation is just calling the `Task`-based API, then additional integration tests for the `Observable`s are just doubling up. But yes, :thumbsup: to additional unit tests.

**Comment by ryangribble** [comment]
Oops!  Looks like just a minor issue there with the "NotNull" test checking empty strings, and the "NonEmpty" test checking nulls!  If you can just fix that up we are good to go

Also given @shiftkey's comment let's ditch the integration test for the observable implementation, since it's just calling the real client anyway.  SORRY!

**Comment by rogertinsley** [comment]
No problem - I've removed the integration test and updated the unit test. 

I've learnt a lot about the codebase and how to contribute with this PR and you guys haven't scared me off either :grin: so I'm going to pick up some more `up-for-grabs` :smile:

**Comment by shiftkey** [comment]
@rogertinsley this all looks good! I can't wait for the next one.

I'll let @ryangribble have the final word.

**Comment by ryangribble** [comment]
> I've learnt a lot about the codebase and how to contribute with this PR and you guys haven't scared me off either

Great to hear!  Bring on the next one, this one LGTM
LGTM
